src/manager.c

```diff
@@ -323,15 +323,43 @@ mapping_free(struct CompMapping *to_unmap)
  * is currently unused. We need to do some reinitialisation (writable segments,
  * scratch memory), but we preclude the need of more memory mapping / unmapping
  * operations, which seem expensive.
+ *
+ * Order is slightly important here, as we might zero out some stuff we need
+ * for further zero-ing. Currently, it seems we must ensure the heap
+ * allocations are zeroed before the segments, but there might be further
+ * dependencies in the future
  */
 static void
 mapping_reuse(struct CompMapping *to_reuse)
 {
     void *addr = to_reuse->map_addr;
 
+    // Zero out heap allocations
+    // XXX this doesn't clear the whole heap, so any out-of-bounds writes
+    // inside the compartment can leak information to future compartments
+    if (to_reuse->comp->heap_mem_header)
+    {
+        destroy_heap_allocations(to_reuse->comp->heap_mem_header);
+    }
+
+    // Zero out stack
+    bzero(comp_ptr_to_mapping_addr((char *) to_reuse->comp->scratch_mem_base
+                  + to_reuse->comp->scratch_mem_extra,
+              addr),
+        to_reuse->comp->scratch_mem_stack_size);
+
+    // Copy over extra scratch memory
+    /*size_t b_scratch = bench_init("reuse-scratch");*/
+    /*bench_start(b_scratch);*/
+    memcpy(comp_ptr_to_mapping_addr(to_reuse->comp->scratch_mem_base, addr),
+        comp_ptr_to_mapping_addr(
+            to_reuse->comp->scratch_mem_base, to_reuse->comp->staged_addr),
+        to_reuse->comp->scratch_mem_extra);
+    /*bench_end(b_scratch);*/
+
     // Copy over write-permission segments
-    size_t b_segs = bench_init("reuse-segs");
-    bench_start(b_segs);
+    /*size_t b_segs = bench_init("reuse-segs");*/
```

Comment:
Would it be better to make these macros that you can turn on/off?

Reply:
Whoops, these weren't meant to be in here. Removed in https://github.com/capablevms/CHERI-ELF-comp/pull/45/commits/dc9a599687f7c0ec78c6433d5fab9cac3d24599e. 

I have a macro for a single function call, but these multi-line ones I haven't converted, and I think at this point, they're fine as is.